[Mobile Payments] Handle cancellation from card readers

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8085 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

In sTAP Away, we're adding support for Apple's built in card readers to take in-person payments. We continue to use Stripe's SDK to handle payments.

During this work, we are looking to share code where practical with the bluetooth reader flows, and fix issues in those flows which were less obvious with an external reader. One such issue is cancelation from card readers, including Apple's built in card reader.

On the built-in reader modal which Apple provide, there is a `x` button to cancel the payment. On the WisePad 3 reader, there are customer-facing buttons, and tapping the red `x` button when collecting card details will cancel the payment.

In both these cases, we should end the payment flow. For the WisePad 3 reader, we should inform the merchant that the payment was canceled, so they understand how the external device is impacting the payment flow.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### Using an iPhone XS or newer on iOS 16.0 or newer on a US store

1. Navigate to `Menu > Settings > Experimental features`
2. Turn on `Tap to Pay on iPhone`
3. Navigate to `Menu > Payments > Collect payment`
4. Go through the payment flow, and select `Card` on the payment method screen
5. When asked for a reader type, tap `Tap to Pay on iPhone` and go through the Terms of Service Apple ID linking (if you've not done so before)
6. Tap the `x` on Apple's screen to cancel
7. Observe that the modal is dismissed, and the payment flow behind it is also dismissed
8. Observe that the `card_present_collect_payment_canceled` tracks event is logged as expected

### On a CA store using a physical WisePad 3 reader – new flow

1. Navigate to `Menu > Settings > Experimental features`
2. Turn on `Tap to Pay on iPhone`
3. Navigate to `Menu > Payments > Collect payment`
4. Go through the payment flow, and select `Card` on the payment method screen
5. Wait for the reader to connect and ask for the card to be tapped or inserted
6. Press the red `x` button on the reader
7. Observe that the payment canceled error (see below) is shown
8. Observe that the `card_present_collect_payment_canceled` tracks event is logged as expected

### On a CA store using a physical WisePad 3 reader – existing flow

1. Navigate to `Menu > Settings > Experimental features`
2. Turn off `Tap to Pay on iPhone`
3. Navigate to `Menu > Payments > Collect payment`
4. Go through the payment flow, and select `Card` on the payment method screen
5. Wait for the reader to connect and ask for the card to be tapped or inserted
6. Press the red `x` button on the reader
7. Observe that the payment flow is dismissed
8. Observe that the `card_present_collect_payment_canceled` tracks event is logged as expected

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

### WisePad 3 reader screen
![payment canceled on reader](https://user-images.githubusercontent.com/2472348/206717773-dafd4b67-1ea7-42fd-8d8b-c97f49d5556f.jpg)

### Apple built in reader flow

https://user-images.githubusercontent.com/2472348/206719025-b64e6b25-c6cd-4202-b573-66d50ccd6cd8.mp4



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

joshheald committed:
8085 Handle cancelation from a card reader

Previously (prior to  5233d70b ) we didn’t dismiss any alerts when cancel was tapped on the reader (or built in reader modal) because we ignored the error.

This corrects the comments, warnings, and errors produced (in the StripeCardReaderService) now that we don’t ignore the error any more.

It also stops showing an alert (with a try again button) when the error is recieved in the CollectOrderPaymentUseCase, and instead dismisses the payment alerts.

Now, in the new flow, we handle cancelations from card readers and from buttons in the app equally: we dismiss the alerts and track the cancelation events.

joshheald committed:
8085 Handle cancellation from a reader in legacy flow

joshheald committed:
8085 Inform the merchant about on-reader cancels

If the user cancels the payment on the reader, it can help the merchant to specifically notify them that the payment has not been taken, and they still need to capture payment.

We do not do this with the built-in reader because it’s clear when everything happens on the same device.

joshheald commented:
@jostnes @lmischner: optional review but since you both noticed this, I thought it's worth tagging you here :)

@toupper: no rush, I've set it for 11.7

toupper commented on Hardware/Hardware/CardReader/StripeCardReader/StripeCardReaderService.swift:
```diff
@@ -502,27 +502,26 @@ private extension StripeCardReaderService {
-                        promise(.failure(CardReaderServiceError.paymentCancellation(underlyingError: underlyingError)))
+                        DDLogWarn("💳 Warning: collect payment cancelled \(error)")
+                        /// If we've not used the cancellable in the app, the cancellation must have come from the reader
+                        if self?.paymentCancellable != nil {
```
Not a blocker but, is there another way to ascertain that the cancellation came from the reader? Tying it to whether the `paymentCancellable` is nil or not seems frail to me, we could change our implementation (e.g making `paymentCancellable` nil in another way) and it would break this logic.

joshheald replied:
I agree, but there's no way that Stripe mention or that I was able to see. I'll check again but this has implications for the other comment on this PR (if there's no way to find out from Stripe, we won't be confident that it was cancelled on device when we get the error from Stripe)

toupper commented on Hardware/Hardware/CardReader/UnderlyingError.swift:
```diff
@@ -29,6 +29,10 @@ public enum UnderlyingError: Error, Equatable {
 
+    /// A command was cancelled on the reader.
+    /// Note that this is not produced by Stripe, we have to infer it from commandCancelled.
+    case commandCancelledOnReader
```
Since we already have a case `commandCancelled`, I would collapse these two checks, or at least add further information to the other case. Otherwise, it might be confusing: If `commandCancelledOnReader` is `true`, `commandCancelled` would be `false`, which is not a valid statement; the command was indeed canceled.

```
case commandCancelledOnDevice
case commandCancelledOnReader
```

or 

```
enum CommandCancellationContext {
    case device
    case reader
}
...
case commandCancelled(context: CommandCancellationContext)
```

joshheald replied:
Agreed, but I think that doing this depends on Stripe telling us the context when we get the error, as we need to convert from the Stripe `commandCancelled` before we have the detail of what triggered the cancellation.

While we could take the second suggestion and default to device, converting (as we do now) to reader when we have the context. There's no difference in code with these, but by saying `device` for reader cancellations too early, we're expressing confidence about something which we don't know for sure (`commandCancelled` could be either source)

toupper commented:
@joshheald thanks for the PR! it tests well and the code looks good to me, I just added a couple of suggestions.

joshheald committed:
8085 Combine cancellation reasons in underlyingerror

joshheald committed:
8085 Explicit flag for cancellation source

joshheald commented:
> thanks for the PR! it tests well and the code looks good to me, I just added a couple of suggestions.

Thanks for the review @toupper 😊 

Unfortunately Stripe don't tell us what the cancellation source is, so we have to handle the unknown case. In bc89e2b I've added a cancellation source enum, which lets us sometimes update the cancellation to have the known type.

In bee90e9 I've updated the service to add a flag which we set when a cancellation starts from the app. Then, if it's not set when we get a cancellation from Stripe during the collect step, we assume that it came from the reader.

joshheald committed:
Fix test
